Mathlib/Analysis/CStarAlgebra/CStarVec.lean

```diff
@@ -0,0 +1,340 @@
+replace the uniformity and bornology by the Pi ones when registering the
+`NormedAddCommGroup (CStarVec n A)` instance. See the docstring of the `TopologyAux` section below
+for more details.
+
+-/
+
+open scoped ComplexOrder RightActions Topology Uniformity Bornology
+
+/-- A vector with entries in a C⋆-algebra and a Hilbert C⋆-module structure. -/
+def CStarVec (n : Type*) (A : Type*) := n → A
+
+namespace CStarVec
+
+section basic
+
+variable {n R A : Type*} (v w : CStarVec n A)
+
+instance [Inhabited A] : Inhabited (CStarVec n A) :=
+  ⟨fun _ => default⟩
+
+/-- The equivalence between `n → A` and `CStarVec n A`. -/
+def ofFun : (n → A) ≃ CStarVec n A := Equiv.refl _
+
+@[simp] lemma ofFun_symm_apply {i : n} : ofFun.symm v i = v i := rfl
+
+@[simp] lemma ofFun_apply {i : n} : ofFun v i = v i := rfl
+
+/-The following should not be a `FunLike` instance because then the coercion `⇑` would get
+unfolded to `FunLike.coe` instead of `ofFun.symm`. -/
+instance instCoeFun : CoeFun (CStarVec n A) (fun _ => (_ : n) → A) where
+  coe := ofFun.symm
+
+@[ext] protected theorem ext (h : ∀ i, v i = w i) : v = w := funext h
+
+instance instZero [Zero A] : Zero (CStarVec n A) :=
+  Pi.instZero
+
+instance instNeg [Neg A] : Neg (CStarVec n A) :=
+  Pi.instNeg
+
+instance instAdd [Add A] : Add (CStarVec n A) :=
+  Pi.instAdd
+
+instance instSub [Sub A] : Sub (CStarVec n A) :=
+  Pi.instSub
+
+instance instAddSemigroup [AddSemigroup A] : AddSemigroup (CStarVec n A) :=
+  { instAdd, Pi.addSemigroup with }
+
+instance instAddMonoid [AddMonoid A] : AddMonoid (CStarVec n A) :=
+  { instAdd, instAddSemigroup, instZero, Pi.addMonoid with }
+
+instance instSubNegMonoid [SubNegMonoid A] : SubNegMonoid (CStarVec n A) :=
+  { instAddMonoid, instNeg, instSub, Pi.subNegMonoid with }
+
+instance instAddGroup [AddGroup A] : AddGroup (CStarVec n A) :=
+  { instSubNegMonoid, Pi.addGroup with }
+
+instance instAddCommMonoid [AddCommMonoid A] : AddCommMonoid (CStarVec n A) :=
+  { instAddMonoid, Pi.addCommMonoid with }
+
+instance instAddCommGroup [AddCommGroup A] : AddCommGroup (CStarVec n A) :=
+  { instAddGroup, Pi.addCommGroup with }
+
+variable (R) in
+instance instSMul [SMul R A] : SMul R (CStarVec n A) :=
+  Pi.instSMul
+
+instance instModule [Semiring R] [AddCommMonoid A] [Module R A] : Module R (CStarVec n A) :=
+  { instSMul R, Pi.module n (fun _ => A) R with }
+
+@[simp] theorem zero_apply [Zero A] {i : n} : (0 : CStarVec n A) i = 0 := rfl
+
+@[simp] theorem add_apply [Add A] {i : n} : (v + w) i = v i + w i := rfl
+
+@[simp] theorem sub_apply [Sub A] {i : n} : (v - w) i = v i - w i := rfl
+
+@[simp] theorem smul_apply (c : R) [SMul R A] {i : n} : (c • v) i = c • v i := rfl
+
+@[simp] theorem neg_apply [Neg A] {i : n} : (-v) i = -v i := rfl
+
+theorem finset_sum_fn {ι : Type*} {s : Finset ι} [AddCommMonoid A] {f : ι → CStarVec n A} :
+    ∑ i ∈ s, ofFun (f i) = (fun j => ∑ i ∈ s, f i j) := by
+  rw [Finset.sum_fn]
+  rfl
```

Comment:
What's this `rfl` unfolding?

Reply:
I think there's some defeq abuse here. I have a PR forthcoming which will change the scope of this PR, and it should address this defeq abuse too. The key is to use the coercion to function manually, because otherwise Lean sees a function application, asks itself "is the type of the thing in the function position a function?", finds the answer is "yes", but commits defeq abuse doing so.